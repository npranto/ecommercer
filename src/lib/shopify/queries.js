import { gql } from '@urql/core';

export const GET_PRODUCTS = gql`
	query getProducts($first: Int = 10) {
		products(first: $first) {
			edges {
				node {
					id
					title
					handle
					description
					images(first: 5) {
						edges {
							node {
								src
								altText
								url
							}
						}
					}
				}
			}
		}
	}
`;
