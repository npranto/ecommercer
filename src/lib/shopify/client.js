'use client';

/**
 * This file contains client-side implementations of Shopify functionality.
 * All client-side interactions with the Shopify API should be imported from here.
 * This ensures proper client-side execution and avoids server-side API token exposure.
 *
 * @example
 * // (client components):
 * import { ShopifyProvider, useProductsQuery } from '@/lib/shopify/client'
 */

import { createClient, defaultExchanges, Provider, useQuery } from 'urql';
import { CONFIG } from '@/config';
import { GET_PRODUCTS } from './queries';

const shopifyClient = createClient({
	url: CONFIG.SHOPIFY_STOREFRONT_API_URL,
	fetchOptions: () => ({
		headers: {
			'X-Shopify-Storefront-Access-Token': CONFIG.SHOPIFY_STOREFRONT_ACCESS_TOKEN,
			'Content-Type': 'application/json',
		},
	}),
	exchanges: [...defaultExchanges],
});

export function ShopifyClientProvider({ children }) {
	return <Provider value={shopifyClient}>{children}</Provider>;
}

export const useProductsQuery = ({ first = 10 }) => {
	const [{ data, fetching, error }] = useQuery({
		query: GET_PRODUCTS,
		variables: {
			first, // number of products to fetch
		},
		requestPolicy: 'cache-and-network', // caches data and only re-fetches when stale
		staleTime: 1000 * 60 * 5, // stale time in milliseconds (default: 5 minute)
	});

	return {
		products: data?.products?.edges || [],
		isLoading: fetching || false,
		error: error || null,
	};
};
