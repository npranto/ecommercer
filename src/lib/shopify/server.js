import { createClient, defaultExchanges } from '@urql/core';
import { CONFIG } from '@/config';
import { GET_PRODUCTS } from './queries';

/**
 * This file contains server-side implementations of Shopify functionality.
 * All server-side interactions with the Shopify API should be imported from here.
 * This ensures proper server-side execution and avoids client-side API token exposure.
 *
 * @example
 * // (server components / API routes):
 * import { getProducts } from '@/lib/shopify/server'
 */

export function createUrqlPromise(client) {
	return async function ({ operation, gql, variables }) {
		if (operation === 'mutation') return await client.mutation(gql, variables).toPromise();
		return await client.query(gql, variables).toPromise();
	};
}

const shopifyClient = createClient({
	url: CONFIG.SHOPIFY_STOREFRONT_API_URL,
	requestPolicy: 'cache-and-network',
	fetchOptions: () => ({
		headers: {
			'X-Shopify-Storefront-Access-Token': CONFIG.SHOPIFY_STOREFRONT_ACCESS_TOKEN,
			'Content-Type': 'application/json',
		},
	}),
	exchanges: defaultExchanges,
});

const shopifyServerQuery = createUrqlPromise(shopifyClient);

export const getProducts = async ({
	first = 10, // number of products to fetch
}) => {
	const { data, error } = await shopifyServerQuery({
		gql: GET_PRODUCTS,
		variables: { first },
	});

	if (error) {
		console.error(error);
		return {
			products: [],
			error,
			isLoading: false,
		};
	}

	return {
		products: data?.products?.edges || [],
		error,
		isLoading: false,
	};
};
