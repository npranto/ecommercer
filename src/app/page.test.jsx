import '@testing-library/jest-dom';
import { render } from '@testing-library/react';
import Page from '@/app/page';
import { getProducts } from '@/lib/shopify/server';

jest.mock('@/lib/shopify/server');

describe('Home Page', () => {
	it('renders loading state', async () => {
		getProducts.mockResolvedValueOnce({
			products: [],
			isLoading: true,
			error: null,
		});

		const { container } = render(await Page());
		expect(container).toMatchSnapshot();
	});
});
