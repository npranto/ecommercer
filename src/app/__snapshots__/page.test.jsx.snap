// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Page renders loading state 1`] = `
<div>
  <p>
    Loading...
  </p>
</div>
`;
